File: network.py
#!/usr/bin/python

import socket, struct, dpkt, pcap
import subprocess, re, jsonpickle
import multiprocessing
import dns, dns.reversename, dns.resolver
import time

'''
 Tasks:

 - monitor pcap for SYN, RST, retransmits, etc. to determine tcp connections
 - for each connection:
   - send CONNECTION [OPEN|ACK|ERROR|CLOSE] to client for all state changes
   - (async) perform dns lookup for host ip, send HOST RESOLVE to client
   - (async) traceroute host to determine path, send HOST TRACE to client for each hop as we receive them
   - (async) whois ip to determine network owner, send NETWORK RESOLVE to client

 Basic actions/protocol:

  CONNECTION OPEN <src> <dst>
  CONNECTION ACK <src> <dst>
  CONNECTION ERROR <src> <dst>
  CONNECTION CLOSE <src> <dst>

  HOST RESOLVE <ip> [<hostname>]
  HOST TRACE <ip> [<hop> <hopip> <ping>]

  NETWORK resolve <ip> [<network> <prefix> <owner>]

'''

'''
Simple base class to allow for easy JSON serialization
'''
class NetworkBase:
  def __str__(self):
    return jsonpickle.encode(self, False)

'''
Classes to represent various levels of a network, as seen from this host
'''
class NetworkDevice(NetworkBase):
  def __init__(self, name):
    self.name = name
    self.interfaces = {}
    self.addresses = {}

  def add_interface(self, ifname):
    self.interfaces[ifname] = NetworkInterface(ifname)

  def add_address(self, ip, prefix, ifname):
    self.addresses[ip] = NetworkAddress(ip, prefix)

class NetworkInterface(NetworkBase):
  def __init__(self, name, mac=None):
    self.name = name
    self.mac = mac

  def set_link(self, type, mac, brd):
    self.mac = mac

class NetworkAddress(NetworkBase):
  hostname = None

  def __init__(self, ip, prefix):
    self.ip = ip
    self.prefix = prefix

'''
Utility class to abstract platform-specific network information gathering
'''
class LinuxNetwork:
  hostname = None

  def get_hostname(self):
    with open('/etc/hostname', 'r') as f:
      return f.read().rstrip()

  def get_host_device(self):
    cmdproc = subprocess.Popen(["ip", "addr", "show"], stdout=subprocess.PIPE)
    cmdout, cmderr = cmdproc.communicate()
    lines = cmdout.split('\n')
    currentiface = None
    device = NetworkDevice(self.get_hostname())

    re_iface = re.compile('^\d+: (.*?)(?:@.*?)?: <(.*?)> (.*)$')
    re_link = re.compile('^\s+link\/(.*?) ([0-9a-f:]{17}) brd ([0-9a-f:]{17})')
    re_addr = re.compile('^\s+inet ([\d\.]+)\/(\d+) (.*)$')

    for i in range(0, len(lines)):
      match = re_iface.match(lines[i])
      if match:
        currentiface = match.group(1)
        device.add_interface(match.group(1))
        continue
      match = re_link.match(lines[i])
      if match:
        device.interfaces[currentiface].set_link(match.group(1), match.group(2), match.group(3))
        continue
      match = re_addr.match(lines[i])
      if match:
        device.add_address(match.group(1), match.group(2), currentiface)
        continue

    return device

def lookuphost(inqueue, outqueue):
  addr = inqueue.get()
  #print "lookup %s" % (addr)
  hostinfo = dns.reversename.from_address(addr)
  hostname = None
  try:
    hostinfo = socket.gethostbyaddr(addr)
    hostname = hostinfo[0]
    #answer = dns.resolver.query(hostinfo, 'PTR')
    #print "got it: %s %s" % (answer.qname, answer.rrset)
    #hostname = str(answer.rrset)
  except (dns.resolver.NXDOMAIN, socket.herror):
    #print "%s not found" % (addr)
    pass
  outqueue.put([addr, hostname])
  return

class AsyncReverseResolver:
  def __init__(self):
    self.manager = multiprocessing.Manager()
    self.inqueue = self.manager.Queue()
    self.outqueue = self.manager.Queue()
    #self.pool = multiprocessing.Pool(processes=100)
    self.procs = []

  def queue(self, addr):
    self.inqueue.put(addr)
    #self.pool.apply_async(lookuphost, [self.inqueue, self.outqueue])
    proc = multiprocessing.Process(target=lookuphost, args=(self.inqueue, self.outqueue))
    proc.daemon = False
    proc.start()
    self.procs.append(proc)

  def begin(self):
    while not self.inqueue.empty():
      self.lookup_next()
  def join(self):
    for i in range(0,len(self.procs)):
      if resolve.procs[i].join(0):
        print "join dat shit: %d" % (i)

  def lookup_next(self, inqueue, outqueue):
    '''
    addr = inqueue.get()
    print "lookup %s" % (addr)
    try:
      hostinfo = socket.gethostbyaddr(addr)
      print hostinfo
      outqueue.put(hostinfo)
    except:
      pass
    '''
    print "hellz yeah"

def ip2str(ip):
  #return struct.pack('L',long(ip))
  return socket.inet_ntoa(ip)

'''
Experimental routine to capture SYN packets from pcap and report all new connections
This is where all the magic is supposed to happen (build a list of hosts, trigger 
async lookups, start traceroutes, etc)
'''
def pcap_capture():
  pc = pcap.pcap()
  pc.setfilter("tcp[13] & 2 == 2")

  decode = { pcap.DLT_LOOP:dpkt.loopback.Loopback,
             pcap.DLT_NULL:dpkt.loopback.Loopback,
             pcap.DLT_EN10MB:dpkt.ethernet.Ethernet }[pc.datalink()]

  for timestamp, packet in pc:
    #print dpkt.ethernet.Ethernet(packet)
    pkt = decode(packet)
    print "[%f] %s:%d => %s:%d" % (timestamp, ip2str(pkt.data.src), pkt.data.data.sport, ip2str(pkt.data.dst), pkt.data.data.dport)

network = LinuxNetwork()
device = network.get_host_device()
print device
pcap_capture()

if False:
  # test code for async resolver
  import random
  resolve = AsyncReverseResolver()
  #resolve.begin()
  #while not (resolve.inqueue.empty() and resolve.outqueue.empty() and resolve.inprogress == 0):
  while True:
    resolve.join()

    if not resolve.outqueue.empty():
      fuh = resolve.outqueue.get()
      print "durf:", fuh

    if random.randint(1,1000) == 1:
      for i in range(0,random.randint(1,5)):
        resolve.queue("64.124.%d.%d" % (random.randint(1,255), random.randint(1,255)))

File: network.php
<?php

class Component_network extends Component {
  public function init() {
    OrmManager::LoadModel("network");
  }

  public function controller_network($args) {
    $vars = array();

    $vars["data"]["device"] = LinuxNetwork::get_host_device();
    $vars["data"]["networks"] = LinuxNetwork::get_networks();
    $vars["data"]["remote"] = $_SERVER["REMOTE_ADDR"];

    return $this->GetComponentResponse("./network.tpl", $vars);
  }
  /**
   * Perform DNS lookups on the provided list of hosts
   */
  public function controller_lookup($args) {
    if (!empty($args["addrs"])) {
      $addrs = explode(",", $args["addrs"]);
      foreach ($addrs as $addr) {
        try {
          $host = @gethostbyaddr($addr);
          if ($host != $addr) {
            $vars["hosts"][$addr] = $host;
          } else {
            $vars["hosts"][$addr] = false;
          }
        } catch (Exception $e) {
          $vars["hosts"][$addr] = false;
        }
      }
    }
    return $this->GetComponentResponse("./lookup.tpl", $vars);
  }
  /**
   * Perform a traceroute to the provided host
   */
  public function controller_trace($args) {
    if (!empty($args["host"])) {
      $host = $args["host"];
      $type = any($args["type"], "udp");
      switch ($type) {
        case 'tcp':
          $vars["hops"] = LinuxNetwork::trace_tcp($host);
          break;
        case 'udp':
          $vars["hops"] = LinuxNetwork::trace_udp($host);
          break;
        case 'icmp':
          $vars["hops"] = LinuxNetwork::trace_icmp($host);
          break;
      }
    }
    return $this->GetComponentResponse("./trace.tpl", $vars);
  }
  public function controller_subnet($args) {
    $ip = $args["ip"];
    $url = "http://whois.arin.net/rest/nets;q={$ip}?showDetails=true&showARIN=false";

    $opts = array(
      'http'=>array(
        'method'=>"GET",
        'header'=>"Accept: application/json\r\n"
      )
    );

    $context = stream_context_create($opts);
    $file = file_get_contents($url, false, $context);
    $data = json_decode($file);
    $net = (is_array($data->nets->net) ? $data->nets->net[0] : $data->nets->net);
    $netblock = (is_array($net->netBlocks->netBlock) ? $net->netBlocks->netBlock[0] : $net->netBlocks->netBlock);
    $subnet["address"] = $netblock->startAddress->{'$'};
    $subnet["prefix"] = $netblock->cidrLength->{'$'};
    $subnet["owner"] = $net->orgRef->{'@name'};
    $vars["subnet"] = $subnet;
    return $this->GetComponentResponse("./subnet.tpl", $vars);
  }
  public function controller_self() {
    return $this->GetComponentResponse("./self.tpl", $_SERVER);
  }
}  

class NetworkSubnet {
  public $address;
  public $prefix;

  public $devices;
}
class NetworkDevice {
  public $name;
  public $interfaces;
  public $addresses;

  public function NetworkDevice($name, $interfaces=null) {
    $this->name = $name;
    if ($interfaces !== null) {
      $this->set_interfaces($interfaces);
    }
  }
  public function add_interface($iface, $addresses=null) {
    $this->interfaces[$iface] = new NetworkInterface($iface, $addresses);
  }
  public function set_interfaces($interfaces) {
    $this->interfaces = $interfaces;
  }
  public function add_address($ip, $prefix, $iface=null) {
    if ($iface !== null) {
      $this->interfaces[$iface]->add_address($ip, $prefix);
    }
    $this->addresses[$ip] = new NetworkAddress($ip, $prefix);
  }
}
class NetworkInterface {
  //public $device;
  public $name;
  public $link;
  public $addresses;
  public $stats;

  public function NetworkInterface($name, $addresses=null) {
    $this->name = $name;
  }
  public function set_link($type, $address, $args) {
    $this->link = new NetworkLink($type, $address, $args);
    return $this->link;
  }
  public function add_address($ip, $network, $args=null) {
    $this->addresses[$ip] = new NetworkAddress($ip, $network, $args);
  }
  public function update_stats($stats) {
    $cols = array(
      "recv.bytes", "recv.packets", "recv.err", "recv.drop", "recv.fifo", "recv.frame", "recv.compressed", "recv.multicast",
      "send.bytes", "send.packets", "send.err", "send.drop", "send.fifo", "send.frame", "send.compressed", "send.multicast"
    );

    for ($i = 0; $i < count($stats); $i++) {
      array_set($this->stats, $cols[$i], $stats[$i]);
    }
  }
}
class NetworkLink {
  public $type;
  public $mac;
  public $capacity;

  public function NetworkLink($type, $mac, $broadcast) {
    $this->type = $type;
    $this->mac = $mac;
  }
}
class NetworkAddress {
  public $ip;
  public $prefix;

  public function NetworkAddress($ip, $prefix, $args=null) {
    $this->ip = $ip;
    $this->prefix = $prefix;
  }
}

class LinuxNetwork {
  public static $hostdevice = null;

  public static function get_hostname() {
    $hostname = trim(file_get_contents("/etc/hostname"));
    return $hostname;
  }
  public static function get_host_device() {
    if (!self::$hostdevice) {
      $device = new NetworkDevice(self::get_hostname());
      $cmd = "ip addr show";
      $cmdout = `$cmd`;
      $lines = explode("\n", $cmdout);
      $ifaces = array();
      $currentiface = null;
      for ($i = 0; $i < count($lines); $i++) {
        if (preg_match("/^\d+: (.*?)(?:@.*?)?: <(.*?)> (.*)$/", $lines[$i], $m)) {
          $currentiface = $m[1];
          $device->add_interface($m[1]);
        } else if (preg_match("/^\s+link\/(.*?) ([0-9a-f:]{17}) brd ([0-9a-f:]{17})/", $lines[$i], $m)) {
          $device->interfaces[$currentiface]->set_link($m[1], $m[2], $m[3]);
        } else if (preg_match("/^\s+inet ([\d\.]+)\/(\d+) (.*)$/", $lines[$i], $m)) {
          $device->add_address($m[1], $m[2], $currentiface);
        }
      }

      self::get_interface_stats($device->interfaces);
      self::$hostdevice = $device;
    }

    return self::$hostdevice;
  }
  public static function get_interface_stats($ifaces) {
    $file = "/proc/net/dev";
    $contents = file($file);
    for ($i = 2; $i < count($contents); $i++) {
      $parts = explode("\t", trim(preg_replace("/\s+/", "\t", $contents[$i])));
      $namepart = array_shift($parts);
      $ifname = substr($namepart, 0, -1);
      if (!empty($ifaces[$ifname]) && $ifaces[$ifname] instanceof NetworkInterface) {
        $ifaces[$ifname]->update_stats($parts);
      }
    }
  }
  public static function get_networks() {
    $cmd = "ip route show";
    $cmdout = `$cmd`;
    $lines = explode("\n", $cmdout);
  }
  public static function get_route_info($addr) {
    $addr = gethostbyname($addr); // make sure we've got an address, not a hostname
    $cmd = "ip route get $addr";
    $cmdout = `$cmd`;
    $lines = explode("\n", $cmdout);
    $info = array();
    if (substr($lines[0], 0, strlen($addr)) == $addr) {
      $parts = explode(" ", substr($lines[0], strlen($addr) + 1));
      for ($i = 0; $i < count($parts) - 1; $i++) {
        if (empty($parts[$i])) continue;
        $info[$parts[$i]] = $parts[++$i];
      }
    }

    return $info;
  }

  /* various types of traceroute */
  public static function trace_icmp($host) {
    return self::trace_mtr($host, false);
  }
  public static function trace_udp($host) {
    return self::trace_mtr($host, true);
  }
  public static function trace_tcp($host) {
    return self::trace_traceroute($host, true);
  }

  protected function trace_traceroute($host, $tcp=false) {
    // old-style traceroute
    $hops = array();

    // Inject our own src address as the first hop for completeness
    $routeinfo = self::get_route_info($host);
    $hops[$host][0] = array($routeinfo["src"], 0);

    if ($tcp) {
      $cmd = "tcptraceroute -w 2 -q 1 -n $host";
    } else {
      $cmd = "traceroute -w 2 -q 1 -n $host |tail -n +2";
    }
    $awkcmd = 'awk \'{print $1 "," $2 "," $3 "," $4}\'';
    $lines = explode("\n", `$cmd | $awkcmd`);
    foreach ($lines as $line) {
      $line = trim($line);
      if (!empty($line)) {
        list($num,$router,$latency,$extra) = explode(",", trim($line));
        if ($latency == "[open]") $latency = $extra;
        if ($router != "*") {
          $hops[$host][(int)$num] = array($router, $latency);
        }
      }
    }
    return $hops;
  }
  protected function trace_mtr($host, $udp=false) {
    $hops = array();
    // Inject our own src address as the first hop for completeness
    $routeinfo = self::get_route_info($host);
    $hops[$host][0] = array($routeinfo["src"], 0);

    $tracecmd = "mtr -c 1 -n --raw $host";
    if ($udp) $tracecmd .= " -u";

    $lines = explode("\n", `$tracecmd`);
    foreach ($lines as $line) {
      $parts = explode(" ", trim($line));
      switch ($parts[0]) {
        case 'h':
          $hops[$host][$parts[1]+1][0] = $parts[2];
          break;
        case 'p':
          $hops[$host][$parts[1]+1][1] = $parts[2];
          break;
      }
    }
    return $hops;
  }
}
